plugins/logs/mask.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2020 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package logs
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/open-policy-agent/opa/util"
+)
+
+type maskOP string
+
+const (
+	maskOPRemove maskOP = "remove"
+	maskOPUpsert maskOP = "upsert"
+)
+
+type maskRule struct {
+	OP           maskOP      `json:"op"`
+	Path         string      `json:"path"`
+	Value        interface{} `json:"value"`
+	escapedParts []string
+}
+
+func (r maskRule) String() string {
+	return "/" + strings.Join(r.escapedParts, "/")
+}
+
+func newMaskRule(op maskOP, path string, val interface{}) (maskRule, error) {
+	rule := maskRule{}
+
+	if len(path) == 0 {
+		return rule, fmt.Errorf("mask must be non-empty")
+	} else if path[0] != '/' {
+		return rule, fmt.Errorf("mask must be slash-prefixed")
+	}
+
+	parts := strings.Split(path[1:], "/")
+
+	if parts[0] != "input" && parts[0] != "result" {
+		return rule, fmt.Errorf("mask prefix not allowed")
+	}
+
+	escapedParts := make([]string, len(parts))
+	for i := range parts {
+		_, err := url.QueryUnescape(parts[i])
+		if err != nil {
+			return rule, err
+		}
+
+		escapedParts[i] = url.QueryEscape(parts[i])
+	}
+
+	rule = maskRule{
+		OP:           op,
+		Path:         path,
+		Value:        val,
+		escapedParts: escapedParts,
+	}
+
+	return rule, nil
+}
+
+func (r maskRule) Mask(event *EventV1) error {
+
+	if len(r.escapedParts) == 1 {
+		switch p := r.escapedParts[0]; p {
+		case "input":
+			switch r.OP {
+			case maskOPRemove:
+				event.Input = nil
+				event.Erased = append(event.Erased, r.String())
+
+			case maskOPUpsert:
+				event.Input = &r.Value
+				event.Masked = append(event.Masked, r.String())
+
+			default:
+				return fmt.Errorf("illegal mask op value: %s", r.OP)
+			}
+
+		case "result":
+			switch r.OP {
+			case maskOPRemove:
+				event.Result = nil
+				event.Erased = append(event.Erased, r.String())
+
+			case maskOPUpsert:
+				event.Result = &r.Value
+				event.Masked = append(event.Masked, r.String())
+
+			default:
+				return fmt.Errorf("illegal mask op value: %s", r.OP)
+			}
+		default:
+			return fmt.Errorf("illegal path value value: %s", p)
+		}
+	} else {
+		var node interface{}
+
+		switch p := r.escapedParts[0]; p {
+		case "input":
+			if event.Input != nil {
+				node = *event.Input
+			}
+		case "result":
+			if event.Result != nil {
+				node = *event.Result
+			}
+		default:
+			return fmt.Errorf("illegal path value value: %s", p)
+		}
+
+		parent := r.lookup(r.escapedParts[1:len(r.escapedParts)-1], node)
```

Comment:
@dkiser the only important comment I have right now is that we should make upsert create intermediate nodes as necessary (like `mkdir -p`). The reason is that otherwise users aren't going to be able to do so reliably because:

* Each mask rule can only generate one operation
* Mask rules are generating a set of operations which are unordered

We could make each make rule generate multiple operations however it'll be somewhat complicated for policies to generate the set of upsert (or other operation types) required to create the hierarchy.

Having the upsert operation create the intermediate nodes resolves these problems.

Reply:
@tsandall  - How should we treat paths with an array index?  Should we ignore these and only allow upsert of objects?

I'll have to mull over how to dynamically created the intermediate nodes.

Reply:
Let's just assume objects for now.

As far as how to create the nodes, here's some code (warning, I haven't tested it):

```golang
func mkdirp(node map[string]interface{}, path []string, value interface{}) bool {
  if len(path) == 0 { return false }
  for i := 0; i < len(path)-1; i++ {
    child, ok := node[path[i]]
    if !ok {
      child := map[string]interface{}{}
      node[path[i]] = child
      node = child
    } else if asobj, ok := child.(map[string]interface{}); ok {
       node = asob
    } else {
      return false
   }
  }
  node[path[len(path)-1]] = value
  return true
}
```

Reply:
Thanks! Playing with these ideas now.

Reply:
@tsandall Added a similar method.  It does make we wonder if we handle `[]interface{}` here as well.